[WIP] Replace Dbghelp with DIA

Warepire opened the pull request:
This is a work in progress, the goal is to get rid of the DbgHelp library and the extendedtrace code. This is 3rd party code that's not really well adapted for our needs. Let's replace this with code that is more adapted to C++ idioms, and that is easily extendable as our needs grow.

The purpose, besides keeping the extendedtrace functionality, is to:
- Make it "easy" to implement a code viewer and tracer of the DLL in Hourglass so that we can debug ourselves, currently we can only really printf-debug the DLL, which is horrible.
- Allow to throw out the trustedcaller-hack in the DLL and make it an debugger-command which can do proper evaluation based on a stack trace.

The latter will cost speed, but the hack is really shitty, and it has to go, there's no way around that, as the hack requires coder-prediction of the call chain, which makes it very hard to properly implement some of the checks in a clean and easy-to-understand manner.

WIP: Please comment more on overall design and look for thinking-failures on my part. This is in too early stages for it to be worth pointing out variable naming and line length problems.

Note: DIA requires unicode to prevent all interactions with the library to be super-painful. I migrated to unicode haphazardly throughout the process of adding DIA, so it's really baked into everywhere. Sorry =/

Known issues:
- PE header parsing is broken for some DLLs, also, the code should move.
- The interface to the debug core is incomplete, it misses everything that is useful.

Warepire committed:
WIP - Begin DbgHelp wrapper to obsolete extendedtrace

This is a work in progress, the goal is to get rid of the DbgHelp library and the extendedtrace code. This is 3rd party code that's not really well adapted for our needs. Let's replace this with code that is more adapted to C++ idioms, and that is easily extendable as our needs grow.

The purpose, besides keeping the extendedtrace functionality, is to:

*  Make it "easy" to implement a code viewer and tracer of the DLL in Hourglass so that we can debug ourselves, currently we can only really printf-debug the DLL, which is horrible.
*  Allow to throw out the trustedcaller-hack in the DLL and make it an debugger-command which can do proper evaluation based on a stack trace.

The latter will cost speed, but the hack is really shitty, and it has to go, there's no way around that, as the hack requires coder-prediction of the call chain, which makes it very hard to properly implement some of the checks in a clean and easy-to-understand manner.

Note: DIA requires unicode to prevent all interactions with the library to be super-painful. I migrated to unicode haphazardly throughout the process of adding DIA, so it's really baked into everywhere. Sorry =/

Known issues:

*  PE header parsing is broken for some DLLs, also, the code should move.
*  The interface to the debug core is incomplete, it misses everything that is useful.

YaLTeR commented on source/application/DbgHelp/DbgHelp.h:
```diff
@@ -0,0 +1,37 @@
+    DbgHelp();
+    ~DbgHelp();
+
+    void AddProcess(HANDLE process, DWORD process_id);
```
Why does this take a process handle along with a pid? Perhaps it’d be better to get the process handle within `AddProcess`? It would also make it more consistent with other functions here which only take a pid.

Warepire replied:
Slow reply, but finally looked into this. You want to create DbgHelp only once, and before any debugged processes have started. This means that you need (a) to handle more than one handle, and (2), the handles aren't available at the time of DbgHelp creation.

YaLTeR commented on source/application/DbgHelp/DbgHelpPrivate.cpp:
```diff
@@ -0,0 +1,178 @@
+     * DataSource tied to 'virtual_address' by checking the position before.
+     */
+    auto it = m_loaded_modules.lower_bound(virtual_address);
+    if (it == m_loaded_modules.begin() || (--it) == m_loaded_modules.begin())
```
Why is `(--it) == m_loaded_modules.begin()` bad?

YaLTeR commented on source/application/DbgHelp/DbgHelpStackWalkCallback.h:
```diff
@@ -0,0 +1,34 @@
+
+class DbgHelpStackWalkCallbackPrivate;
+
+class DbgHelpStackWalkCallback
```
Why is this class needed if all it does is directly relaying everything to the underlying object?

Warepire replied:
Prevents leaking the DIA headers into the main code, they're heavy and intrusive for the IntelliSense.

YaLTeR commented on source/application/DbgHelp/DbgHelpStackWalkCallbackPrivate.cpp:
```diff
@@ -0,0 +1,254 @@
+    if (m_mod_info->m_module_symbol_session == nullptr)
+    {
+        auto it = m_mod_info->m_module_exports_table.lower_bound(GetProgramCounter());
+        if (it != m_mod_info->m_module_exports_table.begin() && (--it) != m_mod_info->m_module_exports_table.begin())
```
Same question of why `(--it) != m_mod_info->m_module_exports_table.begin()` is undesirable.

YaLTeR commented on source/application/DbgHelp/DbgHelpStackWalkCallbackPrivate.cpp:
```diff
@@ -0,0 +1,254 @@
+    {
+        /*
+         * We don't know the parameter names, nor types, nor the actual count, but we have a guess.
+         * We'll also assume that there's no more than 26 parameters to a function.
```
Where does 26 come from?

Warepire replied:
I should clarify that... it's the number of letters in the alphabet.

YaLTeR replied:
Another question is how useful this can be at all? Perhaps not returning anything in case of unknown parameters is the best option? Maybe some way to manually set the parameter count and types through the UI, so through arguments to this (since if this is ever useful it will imply the person will go look up the actual parameter count and types for the function)?

Otherwise, distance to the next frame could provide a better approximation.

YaLTeR commented on source/application/DbgHelp/DbgHelpStackWalkCallbackPrivate.cpp:
```diff
@@ -0,0 +1,254 @@
+}
+
+// TODO: RFC
+void DbgHelpStackWalkCallbackPrivate::EnumerateParameters()
```
What’s the relation between this and `GetFunctionParameters()`? If they do the same thing they should probably be merged together.

Warepire replied:
I'm fairly sure that's what the TODO means.

YaLTeR commented on source/application/Utils/File.cpp:
```diff
@@ -0,0 +1,357 @@
+    if (move.HighPart == 0)
+    {
+        high_part = nullptr;
+    }
```
What’s going on here?

Warepire replied:
SetFilePointer requires this. If you're going to seek a distance that fits within 32 bits, the `high_part` parameter must be null.

YaLTeR commented on source/application/Utils/File.cpp:
```diff
@@ -0,0 +1,357 @@
+            {
+                memcpy(&reloc, relocations + sizeof(this_reloc) + (r * sizeof(WORD)), sizeof(WORD));
+                /*
+                 * reloc entries that have the top 4 bites zero'd are just padding for alignment as
```
Bytes? Bits? Judging from the check below it’s bits.

Warepire replied:
Bits, need to fix that typo.

YaLTeR commented on source/application/Utils/File.h:
```diff
@@ -0,0 +1,126 @@
+{
+    namespace File
+    {
+        class File
```
So why is a custom WinAPI wrapper needed here and why can’t the standard library facilities be used instead?

Warepire replied:
The idea is that you should not forget to close the file, kinda like `shared_ptr` and `unique_ptr` does it for pointers.

YaLTeR replied:
Why not use `std::ifstream` or wrap a much nicer `FILE*`?

Warepire replied:
I believe I concluded some files will want to set different options not provided through those APIs.

YaLTeR replied:
Any example for the foreseeable future?

Warepire replied:
The DirLocks functionality, which will want to create a temp-file with "delete-on-close".

YaLTeR commented on source/application/application.vcxproj:
```diff
@@ -20,12 +20,12 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'" Label="Configuration">
     <ConfigurationType>Application</ConfigurationType>
-    <CharacterSet>MultiByte</CharacterSet>
+    <CharacterSet>Unicode</CharacterSet>
```
This stupid Microsoft naming... What is MultiByte and what is Unicode?

Warepire replied:
MultiByte is the c-locale, Unicode is UTF-16. Basically, `string` vs `wstring`.

YaLTeR reviewed (commented):
Did you global find and replace some stuff like char->WCHAR at some point? If that’s in some separate commit it's probably better to move it out of this PR.

Warepire commented on source/application/DbgHelp/DbgHelpPrivate.cpp:
```diff
@@ -0,0 +1,178 @@
+    m_should_uninitialize(true)
+{
+    HRESULT rv = CoInitializeEx(nullptr, COINIT_MULTITHREADED | COINIT_SPEED_OVER_MEMORY);
+    if (rv != S_OK || rv != S_FALSE)
```
Memo: This can't be right.

Warepire commented:
I should probably elaborate a little on the design.

`DbgHelp`: The main helper class, the only reason it's a class is so that it can look after a map of unique_ptr. However, it is fully reasonable to just namespace it and keep the map local to the DbgHelp.cpp file. This will likely be done at some point, just to remove the need for the object. It's meant to be created before the debugging session is started, which is why `AddProcess` takes both a `HANDLE` and a pid.

`DbgHelpPrivate`: This class carries most of the power, it's the class that manages all the other `DbgHelp*` classes. Each instance keeps track of a process. So far the functionality is limited to loading local symbols and stackwalking. There is an issue with the DIA interfacing here as I cannot deduce the arch of the executable from the symbols. Would be neat to use that to know how to launch the StackWalkHelper.

`DbgHelpLoadCallback`: Just for some extra debug logging.

`DbgHelpStackWalkHelper`: Helps DIA walk the stack, currently the helper has no support for omit-frame-pointer generated code. Though, it seems really tricky to generate such code. I don't think that support is very urgent. There are some x64 exclusive callbacks that are unimplemented and basically undocumented on MSDN.

`DbgHelpStackWalkCallback`: This class is given to your callback function that you gave in the `StackWalk` call. This allows fetching information about the stack frame.

`DbgHelpStackWalkCallbackPrivate`: Does all the heavy lifting for the callback, it's hiding some caching details from the outside world, and prevents leakage of the DIA headers into the rest of the program, as the headers are really heavy on IntelliSense.

`File` utils: The important part is the `ExecutableFileHeaders` class and the function call `GetExportTable` as it should find all the symbols a DLL exports and at which addresses. This is currently buggy and doesn't work properly with DLLs that have a ordinal-offset. Not sure what the problem is, the code doesn't follow the PE/COFF spec, but following the spec is more broken. The usefulness here is getting symbol names from DLLs we have no PDB for, and later useful for enumerating DLL functions we don't hook.

Warepire commented:
dbghelp.dll can dump the correct export tables from many system DLLs, my code (that follows the spec decently) can't. @YaLTeR would you mind helping me out here? I might need another mind for figuring out how to do it when the DLLs ignore the spec.

YaLTeR commented:
Hm?

Warepire commented:
@YaLTeR  IRC?

YaLTeR commented:
I'll hop on IRC tomorrow (going to sleep right now).

YaLTeR commented:
https://github.com/Warepire/Hourglass-Resurrection/pull/4

YaLTeR commented on source/application/DbgHelp/DbgHelpStackWalkCallback.h:
```diff
@@ -0,0 +1,34 @@
+
+class DbgHelpStackWalkCallbackPrivate;
+
+class DbgHelpStackWalkCallback
```
This having almost the same name as `DbgHelp::StackWalkCallback` yet being completely different things is extremely confusing.

Warepire replied:
That's a very good point.

YaLTeR commented on source/application/DbgHelp/DbgHelpStackWalkCallbackPrivate.cpp:
```diff
@@ -0,0 +1,340 @@
+        if (it != m_mod_info->m_module_exports_table.begin())
+        {
+            it--;
+            function_name = it->second;
```
There should be a check for the function address equality here.

Warepire replied:
But how? We use the contents of the program counter register to find which function we're in, we don't know the length of the function.

YaLTeR replied:
Hm, yeah, I was wondering about this but looking from the following code that queries the symbol I thought the EIP was the address of the function here. What is it then? The return address?

YaLTeR commented on source/application/DbgHelp/DbgHelpStackWalkCallbackPrivate.cpp:
```diff
@@ -0,0 +1,340 @@
+}
+
+// TODO: RFC
+std::wstring DbgHelpStackWalkCallbackPrivate::GetFunctionParameters()
```
I suggest some sort of `std::vector<DbgHelp::Parameter>` where `DbgHelp::Parameter` is essentially a tagged union (think `boost::variant`) for the return type.

Warepire replied:
I like this idea.

YaLTeR commented on source/application/DbgHelp/DbgHelpStackWalkCallbackPrivate.cpp:
```diff
@@ -0,0 +1,340 @@
+        ULONG num_fetched;
+
+        /*
+         * This will be incremented for each found symbol, with it's size.
```
it's => its

YaLTeR commented on source/application/DbgHelp/DbgHelpStackWalkCallbackPrivate.cpp:
```diff
@@ -0,0 +1,340 @@
+            ULONGLONG length = 0;
+            if (type_info->get_length(&length) != S_OK)
+            {
+                sym_info->Release();
```
All this stuff just asks for a wrapper with `sym_info->Release()` in the destructor.

Warepire replied:
Sadly, these throws exceptions when you stuff them in `CComPtr` classes. Because when you reach the end of a DIA enum `sym_info` contains invalid data.

Warepire commented:
The most recent commits takes care of the discussion we had about the alignment and the use of `DWORD32`. It's now switched to using `BYTE` as we agreed upon.

I also converted the DbgHelp class to a namespace, and turned `DbgHelpStackWalkCallback` into an I-class, this removed the need for the class `DbgHelpStackWalkCallbackPrivate` to be named as such, so the Private part was dropped.

YaLTeR commented on source/application/Utils/File.cpp:
```diff
@@ -186,12 +186,12 @@ Utils::File::ExecutableFileHeaders::ExecutableFileHeaders(const std::wstring& fi
 
 
-DWORD Utils::File::ExecutableFileHeaders::GetImageSizeInRAM() const
+DWORD Utils::File::ExecutableFileHeaders::GetImageSizeInRAM()
```
Why was `const` removed here and in some places below?

Warepire replied:
This is because the code now accesses the buffer directly. There's a lot of stupid `const_cast` necessary otherwise, because I can't cast the pointer. If you have a nicer way to do it, please show me.

YaLTeR replied:
How about `&m_buffer[offset]`?

YaLTeR commented on source/application/DbgHelp/DbgHelp.h:
```diff
@@ -20,6 +20,6 @@ namespace DbgHelp
     void LoadSymbols(DWORD process_id, HANDLE module_file, LPCWSTR module_name, DWORD64 module_base);
 
-    using StackWalkCallback = std::function<DbgHelpStackWalkCallback::Action(const DbgHelpStackWalkCallback&)>;
+    using StackWalkCallback = std::function<IDbgHelpStackWalkCallback::Action(IDbgHelpStackWalkCallback*)>;
```
Why change reference to pointer?

Warepire commented:
Memo to myself: Validate that no `std::wstring::c_str()` strings are passed to BSTR COM methods.

YaLTeR committed:
Mark DbgHelpStackWalkCallback functions with override

YaLTeR commented on source/application/DbgHelp/IDbgHelpStackWalkCallback.h:
```diff
@@ -0,0 +1,29 @@
+    // TODO: Methods
+    virtual const std::wstring GetModuleName() = 0;
+    virtual const std::wstring GetFunctionName() = 0;
+    virtual const DWORD GetParameterCount() = 0;
```
What are `const`s doing here? If they mark constness of functions they should be in the end.

Warepire replied:
They started out as references I think, and then I forgot to remove `const`. That's my best guess.

YaLTeR commented on source/application/DbgHelp/DbgHelpTypes.h:
```diff
@@ -0,0 +1,33 @@
+    Int,
+    UnsignedInt,
+    Long,
+    UnsignedLong,
```
I think it's better to replace this stuff with variants meaning `int8_t`, `uint8_t`, `int16_t`, `uint16_t`, and so on.

Warepire replied:
Yeah, might be a better way to display it.

YaLTeR committed:
Remake the DbgHelp callback function parameter interface

YaLTeR committed:
Fix an off-by-one in TranslateDeviceName()

YaLTeR committed:
Fix pointer type printing

YaLTeR committed:
Change static to anonymous namespace

YaLTeR committed:
Add enum support to stack walk callback

YaLTeR committed:
Rewrite the stack trace function printing

YaLTeR committed:
Display unsure status in stack trace

YaLTeR committed:
Specify the captures in a lambda

YaLTeR committed:
Change printed type names

YaLTeR committed:
Mark visitor classes with #pragma message

YaLTeR committed:
Move default arg names to GetParameters()

YaLTeR committed:
Change i32 to s32

YaLTeR committed:
Add support for void and UDTs

YaLTeR committed:
Filter out anything but function parameters

YaLTeR committed:
Format comments

YaLTeR committed:
Add support for extra char types

YaLTeR committed:
Print extra character types in the stack trace

YaLTeR committed:
Add static asserts for extra char types' sizes

YaLTeR commented:
There's an exception in the `CComPtrBase` destructor consistently happening on exiting Hourglass after having run a program once. Looking at the call stack, it happens inside the `DbgHelpPrivate` destructor called on deinitializing global variables.

YaLTeR committed:
Hack to fix Hourglass not working off of network drives

YaLTeR committed:
Support iterating over DIA enumerators with range-for

YaLTeR committed:
Update code style

Warepire committed:
Resolve a misconception with how the COM library works

Fixes a crash on close because the COM Library is uninitialized too early.

Adds a COM Library singleton class to COM utils, because an application
can only uninitialize it once, and this beats reference tracking.

Warepire commented:
Fixed the exception and checked the strings passing to the COM Library.

Warepire commented:
Known remaining TODO:

- [x] Move stack trace print code to logging.
- [x] Clean up commented out code.
- [x] Consider removing the File class, and move the ExecutableHeaders class to a new file.
- [x] Erase extendedtrace
- [x] Update dia.dll to the version that ships with VS2017

Warepire commented:
@YaLTeR Please look into this warning:
```
2>z:\hourglass-resurrection\source\application\dbghelp\dbghelptypes.h(77): warning C4715: 'DbgHelpBasicType::GetName': not all control paths return a value
2>z:\hourglass-resurrection\source\application\dbghelp\dbghelpstackwalkcallback.cpp(482): warning C4715: '`DbgHelpStackWalkCallback::GetParameterValue'::`2'::visitor::operator()': not all control paths return a value
```

Also, this data looks bogus:
```
Z:\Hourglass-Resurrection\build\Debug\hooks.dll!0x0x10010bc8 : IPC::SendIPCMessage(enum IPC::Command : u64 cmd = 0x2100042a6, void * data = 0x00000000, u32 data_size = 0x49c940)
```

Warepire committed:
Clean up the code a bit and squash some TODOs

Warepire committed:
Delete extended trace and move the stack trace printer to logging

Warepire committed:
Fix a warning in File.cpp

YaLTeR commented:
Before touching anything I launched the current code with the same test exe as I used and it crashes in `PrintStackTrace()`, more precisely in `DbgHelpStackWalkCallback::EnumerateParameters()` on line 366 because `symbol` returned by `GetFunctionSymbol()` is for some reason 0.

YaLTeR commented:
As for the warnings, unfortunately it's [VS being stupid](http://imgur.com/a/BehnZ). I guess I can add something like `default: assert(false); return 0;` to make it shut up.

Warepire commented:
@YaLTeR I do not see any crashes with my test executables. IRC?

Warepire commented:
Leave the warnings be in this case, I agree that it's a stupid warning.

Warepire commented:
@YaLTeR I traced a bit about the crash, this is my current status:

```
Displacement = 0x0
                        SymIndex:        0x1
                          SymTag:        0xA
                            Name:        @ILT+285(_wWinMainCRTStartup)
                   LexicalParent:        000000000051FAB0
                    LocationType:        0x1
                  AddressSection:        0x2
                   AddressOffset:        0x122
          RelativeVirtualAddress:        0x11122
                  VirtualAddress:        0x11122
                          Length:        0x5
                 LexicalParentId:        0x2
                            Code:        true
                        Function:        true
                         Managed:        false
                            MSIL:        false
                 UndecoratedName:        @ILT+285(_wWinMainCRTStartup)
                   TargetSection:        0x2
                    TargetOffset:        0xEC0
    TargetRelativeVirtualAddress:        0x11EC0
            TargetVirtualAddress:        0x11EC0
                  export is DATA:        false
PublicSymbol: [00011122][0002:00000122] @ILT+285(_wWinMainCRTStartup)(@ILT+285(_wWinMainCRTStartup))
```
The reason we can't find the symbol is because it's registered as a `SymTagPublicSymbol`, I have not found how to convert this to a `SymTagFunction`, it has no children.

Here's another `SymTagPublicSymbol` instance of it, it doesn't have any children either:
```
                        SymIndex:        0x1
                          SymTag:        0xA
                            Name:        _wWinMainCRTStartup
                   LexicalParent:        00000000004819F0
                    LocationType:        0x1
                  AddressSection:        0x2
                   AddressOffset:        0xEC0
          RelativeVirtualAddress:        0x11EC0
                  VirtualAddress:        0x11EC0
                          Length:        0xA
                 LexicalParentId:        0x2
                            Code:        true
                        Function:        true
                         Managed:        false
                            MSIL:        false
                 UndecoratedName:        _wWinMainCRTStartup
                  export is DATA:        false
PublicSymbol: [00011EC0][0002:00000EC0] _wWinMainCRTStartup(_wWinMainCRTStartup)
```

YaLTeR committed:
Add missing \ns

YaLTeR committed:
Fix extra 0xs in the stack traces

YaLTeR committed:
Add a workaround for SymTagPublicSymbol in stack traces

YaLTeR committed:
Fix parameter offsets in stack trace

Warepire committed:
Map files to memory when reading their PE/COFF headers

This gets rid off the File-class, it was a premature idea anyway.

Warepire committed:
Clarify a debug log print

YaLTeR committed:
Add support for ptrs to types with unknown size

Warepire committed:
Merge pull request #12 from YaLTeR/dbghelp-rewrite

Add support for ptrs to types with unknown size

vadosnaprimer commented:
Jee, guys, congrats on finally finishing this monster!
